man/spek.1
.TH SPEK-X 1 "2023\-01\-26" "User\[aq]s Guide" "Version 0.9.1"
.SH NAME
.PP
Spek-X \- Acoustic Spectrum Analyser
.SH SYNOPSIS
.PP
\f[C]spek\f[] [\f[I]OPTION\f[] \f[I]\&...\f[]] [\f[I]FILE\f[]] [\f[I]PNG\f[]] [\f[I]WIDTH\f[]] [\f[I]HEIGHT\f[]]
.SH DESCRIPTION
.PP
\f[I]Spek-X\f[] generates a spectrogram for the input audio file.
.SH ARGUMENTS
.TP
\f[I]FILE\f[]
Audio file to be analyzed.
.TP
\f[I]PNG\f[]
Write spectrogram to this PNG file then exit.
.TP
\f[I]WIDTH\f[]
Initial width of the spectrum in pixels.
.TP
\f[I]HEIGHT\f[]
Initial height of the spectrum in pixels.
.SH OPTIONS
.TP
.B \f[C]\-h\f[], \f[C]\-\-help\f[]
Output the help message then quit.
.RS
.RE
.TP
.B \f[C]\-V\f[], \f[C]\-\-version\f[]
Output version information then quit.
.RS
.RE
.SH KEYBINDINGS
.SS Notes
.PP
On macOS use the Command key instead of Ctrl.
.SS Menu
.TP
.B \f[C]Ctrl\-O\f[]
Open a new file.
.RS
.RE
.TP
.B \f[C]Ctrl\-S\f[]
Save the spectrogram as an image file.
.RS
.RE
.TP
.B \f[C]Ctrl\-E\f[]
Show the preferences dialog.
.RS
.RE
.TP
.B \f[C]F1\f[]
Open online manual in the browser.
.RS
.RE
.TP
.B \f[C]Shift\-F1\f[]
Show the about dialog.
.RS
.RE
.SS Spectrogram
.TP
.B \f[C]c\f[], \f[C]C\f[]
Change the audio channel.
.RS
.RE
.TP
.B \f[C]f\f[], \f[C]F\f[]
Change the DFT window function.
.RS
.RE
.TP
.B \f[C]l\f[], \f[C]L\f[]
Change the lower limit of the dynamic range in dBFS.
.RS
.RE
.TP
.B \f[C]p\f[], \f[C]P\f[]
Change the palette.
.RS
.RE
.TP
.B \f[C]s\f[], \f[C]S\f[]
Change the audio stream.
.RS
.RE
.TP
.B \f[C]u\f[], \f[C]U\f[]
Change the upper limit of the dynamic range in dBFS.
.RS
.RE
.TP
.B \f[C]w\f[], \f[C]W\f[]
Change the DFT window size.
.RS
.RE
.SH FILES
.TP
.B \f[I]~/.config/spek/preferences\f[]
The configuration file for \f[I]Spek-X\f[], stored in a simple INI format.
.RS
.RE
.SH AUTHORS
.PP
Alexander Kojevnikov <alexander@kojevnikov.com>.
Other contributors are listed in the LICENCE.md file distributed with
the source code.
.SH DISTRIBUTION
.PP
The latest version of \f[I]Spek-X\f[] may be downloaded from
<https://github.com/MikeWang000000/spek-X>.


src/spek.cc
#include <wx/cmdline.h>
#include <wx/log.h>
#include <wx/socket.h>

#include "spek-artwork.h"
#include "spek-platform.h"
#include "spek-preferences.h"

#include "spek-window.h"

class Spek: public wxApp
{
public:
    Spek() : wxApp(), window(NULL), quit(false) {}

protected:
    virtual bool OnInit();
    virtual int OnRun();
#ifdef OS_OSX
    virtual void MacOpenFiles(const wxArrayString& files);
#endif

private:
    SpekWindow *window;
    wxString path;
    wxString pngpath;
#ifdef OS_OSX
    bool cmdcall = false;
#endif
    bool quit;
};

IMPLEMENT_APP(Spek)

bool Spek::OnInit()
{
    wxLog::SetLogLevel(wxLOG_Error);
    wxInitAllImageHandlers();
    wxSocketBase::Initialize();

    spek_artwork_init();
    spek_platform_init();

    SpekPreferences& prefs = SpekPreferences::get();
    prefs.init();
    long width = prefs.get_window_width();
    long height = prefs.get_window_height();

    static const wxCmdLineEntryDesc desc[] = {{
            wxCMD_LINE_SWITCH,
            "h",
            "help",
            "Show this help message",
            wxCMD_LINE_VAL_NONE,
            wxCMD_LINE_OPTION_HELP,
        }, {
            wxCMD_LINE_SWITCH,
            "V",
            "version",
            "Display the version and exit",
            wxCMD_LINE_VAL_NONE,
            wxCMD_LINE_PARAM_OPTIONAL,
        }, {
            wxCMD_LINE_PARAM,
            NULL,
            NULL,
            "FILE",
            wxCMD_LINE_VAL_STRING,
            wxCMD_LINE_PARAM_OPTIONAL,
        }, {
            wxCMD_LINE_PARAM,
            NULL,
            NULL,
            "PNG",
            wxCMD_LINE_VAL_STRING,
            wxCMD_LINE_PARAM_OPTIONAL
        },{
            wxCMD_LINE_PARAM,
            NULL,
            NULL,
            "WIDTH",
            wxCMD_LINE_VAL_NUMBER,
            wxCMD_LINE_PARAM_OPTIONAL
        },{
            wxCMD_LINE_PARAM,
            NULL,
            NULL,
            "HEIGHT",
            wxCMD_LINE_VAL_NUMBER,
            wxCMD_LINE_PARAM_OPTIONAL
        },
        wxCMD_LINE_DESC_END,
    };

    wxCmdLineParser parser(desc, argc, argv);
    int ret = parser.Parse(true);
    if (ret == 1) {
        return false;
    }
    if (ret == -1) {
        this->quit = true;
        return true;
    }
    if (parser.Found("version")) {
        // TRANSLATORS: the %s is the package version.
        wxPrintf(_("Spek version %s"), PACKAGE_VERSION);
        wxPrintf("\n");
        this->quit = true;
        return true;
    }
    if (parser.GetParamCount()) {
        this->path = parser.GetParam();
#ifdef OS_OSX
        this->cmdcall = true;
#endif
    }
    if (parser.GetParamCount() > 1) {
        wxString pngpath = parser.GetParam(1);
        wxFileName png_path(pngpath);
        png_path.MakeAbsolute();
        wxFileName file_path(this->path);

        if (!(png_path.IsDirWritable() || png_path.IsFileWritable())) {
            wxPrintf(_("PNG \"%s\" is not writable"), pngpath);
#ifndef OS_WIN
            wxPrintf("\n");
#endif
            this->quit = true;
            return false;
        }
        if (png_path.SameAs(file_path)) {
            wxPrintf(_("PNG \"%s\" is same as FILE \"%s\""), pngpath, this->path);
#ifndef OS_WIN
            wxPrintf("\n");
#endif
            this->quit = true;
            return false;
        }

        this->pngpath = pngpath;
    }
    if (parser.GetParamCount() > 3) {
        wxString width_str = parser.GetParam(2);
        wxString height_str = parser.GetParam(3);

        if (!(width_str.ToLong(&width) && height_str.ToLong(&height))) {
            width = prefs.get_window_width();
            height = prefs.get_window_height();
        }
    }

    this->window = new SpekWindow(width, height, this->path, this->pngpath);
    this->window->Show(true);
    SetTopWindow(this->window);
    return true;
}

int Spek::OnRun()
{
    if (quit) {
        return 0;
    }

    return wxApp::OnRun();
}

#ifdef OS_OSX
void Spek::MacOpenFiles(const wxArrayString& files)
{
    if (!this->cmdcall && files.GetCount() == 1) {
        this->window->open(files[0]);
    }
}
#endif
